source/extensions/filters/http/cache/config.cc

```diff
@@ -1,32 +1,56 @@
 #include "source/extensions/filters/http/cache/config.h"
 
+#include "http_cache.h"
 #include "source/extensions/filters/http/cache/cache_filter.h"
 
 namespace Envoy {
 namespace Extensions {
 namespace HttpFilters {
 namespace Cache {
 
-Http::FilterFactoryCb CacheFilterFactory::createFilterFactoryFromProtoTyped(
-    const envoy::extensions::filters::http::cache::v3::CacheConfig& config,
-    const std::string& stats_prefix, Server::Configuration::FactoryContext& context) {
+namespace {
+template <typename FactoryCtx>
+HttpCachePtr getCache(const envoy::extensions::filters::http::cache::v3::CacheConfig& config,
+                      FactoryCtx& ctx) {
   const std::string type{TypeUtil::typeUrlToDescriptorFullName(config.typed_config().type_url())};
-  HttpCacheFactory* const http_cache_factory =
-      Registry::FactoryRegistry<HttpCacheFactory>::getFactoryByType(type);
-  if (http_cache_factory == nullptr) {
+  auto* http_cache_factory = Registry::FactoryRegistry<HttpCacheFactory>::getFactoryByType(type);
+  if (Registry::FactoryRegistry<HttpCacheFactory>::getFactoryByType(type) == nullptr) {
     throw EnvoyException(
         fmt::format("Didn't find a registered implementation for type: '{}'", type));
   }
 
-  auto cache = http_cache_factory->getCache(config, context);
+  auto cache = http_cache_factory->getCache(config, ctx);
+  if (cache == nullptr) {
+    throw EnvoyException(fmt::format("Didn't find HTTP cache for type: '{}', config: '{}'", type,
+                                     config.DebugString()));
+  }
+
+  return cache;
+}
+} // namespace
 
-  return [config, stats_prefix, &context,
-          cache](Http::FilterChainFactoryCallbacks& callbacks) -> void {
-    callbacks.addStreamFilter(std::make_shared<CacheFilter>(config, stats_prefix, context.scope(),
-                                                            context.timeSource(), *cache));
+Http::FilterFactoryCb CacheFilterFactory::createFilterFactoryFromProtoTyped(
+    const envoy::extensions::filters::http::cache::v3::CacheConfig& config,
+    const std::string& stats_prefix, Server::Configuration::FactoryContext& context) {
+
+  auto cache = getCache(config, context);
+
+  return [config, stats_prefix, cache,
+          &context](Http::FilterChainFactoryCallbacks& callbacks) -> void {
+    callbacks.addStreamFilter(std::make_shared<CacheFilter>(
+        stats_prefix, context.scope(), context.timeSource(), CacheFilterConfig(cache, config)));
```

Comment:
CacheFilter's constructor is taking a reference to CacheFilterConfig, and here you're passing in a temporary instance. As soon as this call goes out of scope that reference will be invalid, and the Filter outlives it.

You can (and should) still construct with a reference though - it just needs to be captured and owned by this lambda, rather than a temporary, i.e. `auto cache_filter_config = CacheFilterConfig(getCache(config, context))` instead of `auto cache =...` before the lambda.

OR you can pass it in as a temporary but make sure it's fully consumed during the constructor, i.e. initialize cache_ and vary_allow_headers_ and don't store config_.